src/classes/PS_IntegrationService.cls

```diff
@@ -54,33 +54,145 @@ global class PS_IntegrationService {
+        List<Payment_Services_Configuration__c> existingConfigurations =
+            getExistingRecords(configurationsToUpsertByServiceKey.keySet());
+
+        if (!existingConfigurations.isEmpty()) {
+            applyExistingIdsToNewConfigurations(existingConfigurations, configurationsToUpsertByServiceKey);
+        }
+
+        Set<String> existingConfigurationServiceKeys = collectServiceKeys(existingConfigurations);
+
+        configurationsToInsertByServiceKey = configurationsToUpsertByServiceKey.clone();
+        configurationsToInsertByServiceKey.keySet().removeAll(existingConfigurationServiceKeys);
+
+        configurationsToUpdateByServicekey = configurationsToUpsertByServiceKey.clone();
+        configurationsToUpdateByServiceKey.keySet().retainAll(existingConfigurationServiceKeys);
+
+        insertConfigurationRecords(configurationsToInsertByServiceKey.values());
+        updateConfigurationRecords(configurationsToUpdateByServiceKey.values());
+    }
+
+    /*******************************************************************************************************
+    * @description Creates a map of Payment_Services_Configuration__c records by their Service_Key__c from
+    * the parameters provided by the caller that need to be upserted.
+    *
+    * @param configParams contains settings, such as the baseURL for connecting to payments services
+    * @return A map of Payment_Services_Configuration__c records by their Service_Key__c
+    */
+    private static Map<String, Payment_Services_Configuration__c> createConfigurationsToUpsert(
+        Map<String, List<Map<String, String>>> configParams) {
 
         if (configParams == null || configParams.isEmpty()) {
             throw new IntegrationServiceException(INVALID_CONFIG_PARAMS);
         }
 
-        List<Payment_Services_Configuration__c> configurations = new List<Payment_Services_Configuration__c>();
+        Map<String, Payment_Services_Configuration__c> configurationsToUpsert =
+            new Map<String, Payment_Services_Configuration__c>();
 
         for (String service : configParams.keySet()) {
             for (Map<String, String> valueMap : configParams.get(service)) {
-                configurations.add(createConfiguration(service, valueMap));
+                Payment_Services_Configuration__c configuration = createConfiguration(service, valueMap);
+                configurationsToUpsert.put(configuration.Service_Key__c, configuration);
             }
         }
 
+        return configurationsToUpsert;
+    }
+
+    /*******************************************************************************************************
+    * @description Queries for existing Payment_Services_Configuration__c records
+    *
+    * @param serviceKeys A set of Payment_Services_Configuration__c.Service_Key__c values
+    */
+    private static List<Payment_Services_Configuration__c> getExistingRecords(Set<String> serviceKeys) {
+        return [
+            SELECT Id,
```

Comment:
For future reference - Id is always returned in SOQL queries so doesn't need to be explicitly named in the field set (I forget this often :)

Reply:
It doesn't hurt to include it though - I usually do just because ... I guess I feel anxious if it isn't in the field list even though I know it's always included